app/controllers/declaration/AdditionalFiscalReferencesController.scala

```diff
@@ -0,0 +1,132 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.declaration
+
+import config.AppConfig
+import controllers.actions.{AuthAction, JourneyAction}
+import controllers.util.CacheIdGenerator.goodsItemCacheId
+import controllers.util._
+import controllers.util.MultipleItemsHelper
+import forms.declaration.{AdditionalFiscalReference, AdditionalFiscalReferencesData}
+import forms.declaration.AdditionalFiscalReference.form
+import forms.declaration.AdditionalFiscalReferencesData._
+import handlers.ErrorHandler
+import javax.inject.Inject
+import models.requests.JourneyRequest
+import play.api.data.Form
+import play.api.i18n.I18nSupport
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents, Result}
+import services.countries.Countries
+import services.model.AutoCompleteItem
+import services.{Country, CustomsCacheService}
+import uk.gov.hmrc.play.bootstrap.controller.FrontendController
+import views.html.declaration.additional_fiscal_references
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class AdditionalFiscalReferencesController @Inject()(
+  authenticate: AuthAction,
+  journeyType: JourneyAction,
+  countries: Countries,
+  errorHandler: ErrorHandler,
+  customsCacheService: CustomsCacheService,
+  mcc: MessagesControllerComponents
+)(implicit appConfig: AppConfig, ec: ExecutionContext)
+    extends FrontendController(mcc) with I18nSupport {
+
+  implicit val countryList: List[AutoCompleteItem] = getCountryData(countries.all)
+
+  def getCountryData(countries: List[Country]): List[AutoCompleteItem] =
+    countries.map(country => AutoCompleteItem(country.countryName, country.countryCode))
+
+  def displayPage(): Action[AnyContent] = (authenticate andThen journeyType).async { implicit request =>
+    customsCacheService.fetchAndGetEntry[AdditionalFiscalReferencesData](goodsItemCacheId, formId).map {
+      case Some(data) => Ok(additional_fiscal_references(form, data.references))
+      case _          => Ok(additional_fiscal_references(form))
+    }
+  }
+
+  def saveReferences(): Action[AnyContent] = (authenticate andThen journeyType).async { implicit request =>
+    val actionTypeOpt = request.body.asFormUrlEncoded.map(FormAction.fromUrlEncoded(_))
+
+    val cachedData = customsCacheService
+      .fetchAndGetEntry[AdditionalFiscalReferencesData](goodsItemCacheId, formId)
+      .map(_.getOrElse(AdditionalFiscalReferencesData(Seq.empty)))
+
+    val boundForm = form.bindFromRequest()
+
+    cachedData.flatMap { cache =>
+      actionTypeOpt match {
+        case Some(Add)             => addReference(boundForm, cache)
+        case Some(SaveAndContinue) => saveAndContinue(boundForm, cache)
+        case Some(Remove(values))  => removeReference(values, cache)
+        case _                     => errorHandler.displayErrorPage()
```

Comment:
It's common implementation inside our Controllers, I'll try to found reason why we have `_` here

Reply:
Changed, but I think it needs some discussion with team

Reply:
Came back to `_`. We have warning in console about it:
```
match may not be exhaustive.
[warn] It would fail on the following inputs: None, Some(Continue)
[warn]       actionTypeOpt match {
[warn]       ^
```